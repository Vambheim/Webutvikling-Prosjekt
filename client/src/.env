# Måte å tilby kryptert API-nøkkel (gir tryggere kommunikasjon mot nettet) 
REACT_APP_API_KEY=5b7869e75f75481a8f5005bc7bffd3a8 
