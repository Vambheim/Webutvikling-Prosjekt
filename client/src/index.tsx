import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { Alert, Card } from './widgets';
import {
  RecipeAdd,
  RecipeList,
  ShoppingList,
  TaskDetails,
  TaskEdit,
  TaskNew,
} from './recipe-components';
import Container from 'react-bootstrap/Container';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import NavDropdown from 'react-bootstrap/NavDropdown';
// import Card from 'react-bootstrap/Card';

class Menu extends Component {
  render() {
    return (
      <Navbar bg="light" expand="lg">
        <Container>
          <Navbar.Brand href="/" className="ms-auto">
            Java matapplikasjon
          </Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="ms-auto">
              <Nav.Link href="/#/recipes">Oppskrifter</Nav.Link>
              <Nav.Link href="/#/recipes/add">Legg til oppskrift</Nav.Link>
              <Nav.Link href="/#/recipes/cart">Handleliste</Nav.Link>
              <NavDropdown title="Innstillinger" id="basic-nav-dropdown">
                <NavDropdown.Item href="/#/recipes/cart">Profil</NavDropdown.Item>
                <NavDropdown.Item href="#action/3.2">MÃ¸rk modus</NavDropdown.Item>
                <NavDropdown.Divider />
                <NavDropdown.Item href="#action/3.4">Overrask meg</NavDropdown.Item>
              </NavDropdown>
            </Nav>
          </Navbar.Collapse>
        </Container>
      </Navbar>
    );
  }
}

class Home extends Component {
  render() {
    return <Card title="Welcome">This is your favourite food recipe app</Card>;
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <Alert />
      <Menu />
      <Route exact path="/" component={Home} />
      <Route exact path="/recipes" component={RecipeList} />
      <Route exact path="/recipes/add" component={RecipeAdd} />
      <Route exact path="/recipes/cart" component={ShoppingList} />
      <Route exact path="/tasks/:id(\d+)" component={TaskDetails} /> {/* id must be number */}
      <Route exact path="/tasks/:id(\d+)/edit" component={TaskEdit} /> {/* id must be number */}
      <Route exact path="/tasks/new" component={TaskNew} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);
