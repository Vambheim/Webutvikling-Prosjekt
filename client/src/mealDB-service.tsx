import axios from 'axios';

axios.defaults.baseURL = 'https://www.themealdb.com/api/json/v1/1';

export type MealDBRecipe = {
  mealDBId: number;
  name: string;
  category: string;
  country: string;
};

class MealDBService {
  /**
   * Get task with given id.
   */
  get(mealDBId: number) {
    return axios.get<MealDBRecipe>('/lookup.php?i=' + mealDBId).then((response) => response.data);
  }

  /**
   * Get all tasks.
   */
  // getAll() {
  //   return axios.get<Recipe[]>('/recipes').then((response) => response.data);
  // }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  // create(title: string) {
  //   return axios
  //     .post<{ id: number }>('/tasks', { title: title })
  //     .then((response) => response.data.id);
  // }
}

const mealDBService = new MealDBService();
export default mealDBService;
