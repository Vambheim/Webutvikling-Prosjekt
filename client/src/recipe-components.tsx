import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button } from './widgets';
import { NavLink } from 'react-router-dom';
import recipeService, { Recipe, Step, Ingredient, User } from './recipe-service';
import { createHashHistory } from 'history';

//false as default
export let loggedIn: boolean = false;
const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
/**
 * Renders task list.
 */

export class RecipeList extends Component {
  country: string = '';
  category: string = '';
  recipes: Recipe[] = [];
  filtered_recipes: Recipe[] = [];

  search_input: string = '';

  render() {
    return (
      <>
        <Card title="Filter">
          <Row>
            <Column width={3}>Country:</Column>
            <Column width={3}>Category:</Column>
            <Column width={3}>Ingredients:</Column>
          </Row>
          <Row>
            <Column width={3}>
              <Form.Select
                value={this.country}
                onChange={(event) => (this.country = event.currentTarget.value)}
              >
                {this.recipes
                  .map((recipe) => recipe.country)
                  .filter((country, index, array) => array.indexOf(country) === index)
                  .map((country, i) => (
                    <option key={i} value={country}>
                      {country}
                    </option>
                  ))}
              </Form.Select>
            </Column>
            <Column width={3}>
              <Form.Select
                value={this.category}
                onChange={(event) => (this.category = event.currentTarget.value)}
              >
                {this.recipes
                  .map((recipe) => recipe.category)
                  .filter((category, index, array) => array.indexOf(category) === index)
                  .map((category, i) => (
                    <option key={i} value={category}>
                      {category}
                    </option>
                  ))}
              </Form.Select>
            </Column>
            <Column width={3}>
              {/* <Form.Select
                // Hvordan skal vi gjÃ¸re det med filter knyttet til ingrediens?
                value={this.ingredient}
                onChange={(event) => (this.ingredient = event.currentTarget.value)}
              >
                {this.ingredients.map((ing) => (
                  <option key={ing} value={ing}>
                    {ing}
                  </option>
                ))}
              </Form.Select> */}
            </Column>
            <Column>
              <Button.Success onClick={() => this.filter()}>Add filters</Button.Success>
            </Column>
          </Row>
        </Card>

        <Card title="Search">
          <Column>
            <Form.Input
              onChange={(event) => (this.search_input = event.currentTarget.value)}
              value={this.search_input}
              type="search"
              placeholder="Search"
            ></Form.Input>
          </Column>
          <Column>
            <Button.Light onClick={() => this.search()}>Search</Button.Light>
          </Column>
        </Card>
        <Card title="Recepies">
          {this.recipes.map((recipe) => (
            <Row key={recipe.recipe_id}>
              <Column>
                <NavLink
                  to={'/recipes/' + recipe.recipe_id}
                  style={{
                    color: '#9FC1C0',
                    textDecoration: 'none',
                  }}
                >
                  {recipe.name}
                </NavLink>
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    recipeService
      .getAll()
      .then((recipes) => (this.recipes = recipes))
      .catch((error) => Alert.danger('Error getting recipe: ' + error.message));
  }

  search() {
    this.recipes.map((recipe) => {
      if (this.search_input.length > 0) {
        if (recipe.name.includes(this.search_input)) {
          this.filtered_recipes.push(recipe);
          console.log(recipe.name);
        } else {
          console.log('does not match bro');
        }
      }
    });
  }

  filter() {
    Alert.danger('Not yet implemented');
  }
}

export class RecipeDetails extends Component<{ match: { params: { recipe_id: number } } }> {
  recipe: Recipe = { recipe_id: 0, name: '', category: '', country: '' };
  steps: Step[] = [];
  ingredients: Ingredient[] = [];
  portions: number = 1;

  render() {
    return (
      <>
        <Card title={'Recipe for ' + this.recipe.name}>
          <Row>
            <Column width={2}>Name:</Column>
            <Column>{this.recipe.name}</Column>
          </Row>
          <Row>
            <Column width={2}>Category:</Column>
            <Column width={2}>{this.recipe.category}</Column>
          </Row>
          <Row>
            <Column width={2}>Country:</Column>
            <Column width={2}>{this.recipe.country}</Column>
          </Row>
          <Button.Light onClick={() => this.loggedInCheck()}>
            Like this recipe &#10084;
          </Button.Light>
        </Card>
        <Card title="This is how you make it">
          <ol>
            {this.steps.map((step) => (
              <Row key={step.order_number}>
                <Column>
                  <li>{step.description}</li>
                </Column>
              </Row>
            ))}
          </ol>
        </Card>
        <Card title="Ingredients">
          <Row>
            <Column width={2}>Select portions jajja:</Column>
            <Column width={6}>
              <Form.Input
                type="number"
                value={this.portions}
                onChange={(event) => (this.portions = Number(event.currentTarget.value))}
                min={1}
                max={50}
              ></Form.Input>
            </Column>
          </Row>

          {this.ingredients.map((ing) => (
            <Row key={ing.ingredient_id}>
              <Column width={2}>
                {ing.amount_per_person * this.portions +
                  ' ' +
                  ing.measurement_unit +
                  ' ' +
                  ing.name}
              </Column>
              <Column>
                {/* Adds to shopping list, if logged in */}
                <Button.Light onClick={() => this.loggedInCheck()}>&#128722;</Button.Light>
              </Column>
            </Row>
          ))}
          <Row>
            <Column>
              <Button.Success onClick={() => this.addIngToShoppingList()}>
                Add all ingredients to shopping list
              </Button.Success>
            </Column>
          </Row>
        </Card>

        <Button.Success
          onClick={() => {
            this.editRecipe();
          }}
        >
          Edit
        </Button.Success>
      </>
    );
  }

  mounted() {
    recipeService
      .get(this.props.match.params.recipe_id)
      .then((recipe) => (this.recipe = recipe))
      .then(() => recipeService.getSteps(this.recipe.recipe_id))
      .then((steps) => (this.steps = steps))
      .then(() => recipeService.getIngredients(this.recipe.recipe_id))
      .then((ingredients) => (this.ingredients = ingredients))
      .catch((error) => Alert.danger('Error getting recipe details: ' + error.message));
  }

  loggedInCheck() {
    if (!loggedIn) {
      Alert.info(`You have to log in to like this recipe`);
    }
  }

  editRecipe() {
    history.push('/recipes/' + this.props.match.params.recipe_id + '/edit');
  }

  addIngToShoppingList() {
    Alert.danger('Not yet implemented');
  }
}

export class RecipeAdd extends Component {
  steps: Step[] = [];
  ingredients: Ingredient[] = [];
  recipe: Recipe = { recipe_id: 0, name: '', category: '', country: '' };
  showIng: string = 'hidden';
  showSteps: string = 'hidden';

  render() {
    return (
      <>
        <Card title="Add Recipe">
          <Row>
            <Column>You can add your favourite recipe here</Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Input
                value={this.recipe.name}
                type="text"
                placeholder="Name"
                onChange={(event) => (this.recipe.name = event.currentTarget.value)}
              ></Form.Input>
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Input
                value={this.recipe.category}
                type="text"
                placeholder="Category"
                onChange={(event) => (this.recipe.category = event.currentTarget.value)}
              ></Form.Input>
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Input
                value={this.recipe.country}
                type="text"
                placeholder="Country"
                onChange={(event) => (this.recipe.country = event.currentTarget.value)}
              ></Form.Input>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button.Light onClick={() => (this.showIng = 'visible')}>
                Continue to add ingredients
              </Button.Light>
            </Column>
          </Row>
        </Card>
        <div
          style={{
            //@ts-ignore
            visibility: this.showIng,
          }}
        >
          <Card title="Add ingredients">
            <Row>
              <Column width={2}>
                <Form.Input type="text" placeholder={'Ingredient'}></Form.Input>
              </Column>
              <Column>
                <Button.Light onClick={() => this.addIngredient()}>+</Button.Light>
              </Column>
            </Row>
            <Row>
              <Column>
                <Button.Light onClick={() => (this.showSteps = 'visible')}>
                  Continue to add steps
                </Button.Light>
              </Column>
            </Row>
          </Card>
        </div>
        <div
          style={{
            //@ts-ignore
            visibility: this.showSteps,
          }}
        >
          <Card title="Add steps">
            <Row>
              <Column width={2}>
                <Form.Input placeholder="Step"></Form.Input>
              </Column>
              <Column>
                <Button.Light onClick={() => this.addStep()}>+</Button.Light>
              </Column>
            </Row>
            <Row>
              <Column>
                <Button.Success onClick={() => this.saveRecipe()}>Save recipe</Button.Success>
              </Column>
            </Row>
          </Card>
        </div>
      </>
    );
  }

  saveRecipe() {
    Alert.danger('Not yet implemented');
  }

  addIngredient() {
    Alert.danger('Not yet implemented');
  }

  addStep() {
    Alert.danger('Not yet implemented');
  }
}

export class ShoppingList extends Component {
  ings: string[] = ['Cheese', 'Meat', 'Chicken'];

  render() {
    return (
      <Card title="Shopping List">
        {this.ings.map((ing, i) => (
          <Row key={i}>
            <Row>
              <Column width={3}>{ing}</Column>
              <Column width={1}>
                <Button.Light onClick={() => this.removeOne(i, ing)} small>
                  &#128465;
                </Button.Light>
              </Column>
              <Column></Column>
            </Row>
          </Row>
        ))}
        <Button.Danger
          onClick={() => {
            this.removeAll();
          }}
        >
          Reset Shopping List
        </Button.Danger>
      </Card>
    );
  }

  // mounted() {
  // }

  removeOne(i: number, ing: string) {
    //mÃ¥ gjÃ¸res i database
    if (confirm('Do you want to remove ' + ing + ' from the shopping list?')) {
      // Called when OK is pressed
      this.ings.splice(i, 1);
    } else {
      console.log('Cancel');
    }
  }

  removeAll() {
    //mÃ¥ gjÃ¸res i database
    this.ings = [];
  }
}

export class UserLogIn extends Component {
  username: string = '';
  password: string = '';
  user: User[] = [];

  // autent(username_input: string, password_input: string) {
  //   if (this.users.some((user) => user.username === username_input)) {
  //     console.log('brukenavn godkjent');
  //   } else {
  //     Alert.danger('No user with username: ' + username_input + ' found');
  //   }

  // if (this.users.some((user) => user.password === username_input)) {
  //   console.log('brukenavn godkjent');
  // } else {
  //   Alert.danger('No user with username: ' + username_input + ' found');
  // }

  // console.log('Username: ' + this.email);
  // }

  render() {
    return (
      <Card title="Log In">
        <Row>
          <Column width={6}>
            <Form.Input
              value={this.username}
              type="text"
              placeholder="Username"
              onChange={(event) => (this.username = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column width={6}>
            <Form.Input
              value={this.password}
              type="password"
              placeholder="Password"
              onChange={(event) => (this.password = event.currentTarget.value)}
              // @ts-ignore
              // Makes it possible to log in with enter as well as with button
              onKeyUp={(event) => {
                if (event.key == 'Enter') {
                  console.log(this.username, this.password);
                }
              }}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column width={1}>
            <Button.Success onClick={() => this.logIn()}>Log in</Button.Success>
          </Column>
          <Column width={3}>
            <Button.Light onClick={() => this.clearInput()}>Clear</Button.Light>
          </Column>
        </Row>
        <Row>
          <Column>
            <Button.Light onClick={() => this.createUser()}>No user? Create one here</Button.Light>
          </Column>
        </Row>
      </Card>
    );
  }
  // mounted() {}

  logIn() {
    Alert.danger('Not yet implemented');
  }

  clearInput() {
    this.username = '';
    this.password = '';
  }

  createUser() {
    history.push('/recipes/register');
  }
}

export class RegisterUser extends Component {
  confirm_password: string = '';

  user: User = { username: '', first_name: '', last_name: '', password: '', email: '' };

  render() {
    return (
      <Card title="Create your user here">
        <Row>
          <Column>
            <Form.Input
              type="text"
              value={this.user.username}
              placeholder="Username"
              onChange={(event) => (this.user.username = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Form.Input
              type="text"
              value={this.user.email}
              placeholder="Email"
              onChange={(event) => (this.user.email = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Form.Input
              type="text"
              value={this.user.first_name}
              placeholder="First name"
              onChange={(event) => (this.user.first_name = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Form.Input
              type="text"
              value={this.user.last_name}
              placeholder="Last name"
              onChange={(event) => (this.user.last_name = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Form.Input
              type="password"
              value={this.user.password}
              placeholder="Password"
              onChange={(event) => (this.user.password = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Form.Input
              type="password"
              value={this.confirm_password}
              placeholder="Confirm password"
              onChange={(event) => (this.confirm_password = event.currentTarget.value)}
            ></Form.Input>
          </Column>
        </Row>
        <Row>
          <Column>
            <Button.Success onClick={() => this.createUser()}>Create user</Button.Success>
          </Column>
        </Row>
        <Row>
          <Column>
            <Button.Light onClick={() => this.clearInput()}>Clear</Button.Light>
          </Column>
        </Row>
      </Card>
    );
  }

  createUser() {
    Alert.danger('Not yet implemented');
    // Psuedocode
    // if(nytt_brukernavn finnes i username[]) {
    //   alert("brukernavn finnes")
    // } if (epost finnes) {
    //   alert("epost allerede i bruk, log inn")
    // } if (!password matches) {
    // alert("passord matcher ikke")
    // } else {
    //   let hash = bcrypt.hash(passord);
    //   RecipeService.addUser(this.bruker[ passord= hash])
    // }
  }

  clearInput() {
    this.user = { username: '', first_name: '', last_name: '', password: '', email: '' };
    this.confirm_password = '';
  }
}

export class UserDetails extends Component<{ match: { params: { email: string } } }> {
  render() {
    return (
      <Card title="User details">
        <Row>
          <Column>Username:</Column>
        </Row>
      </Card>
    );
  }

  // mounted() {
  // }
}
//jalla

/**
 * Renders form to edit a specific task.
 */
export class RecipeEdit extends Component<{ match: { params: { id: number } } }> {
  recipe: Recipe = { recipe_id: 0, name: '', category: '', country: '' };
  recipes: Recipe[] = [];

  //test om branch funker
  render() {
    return (
      <>
        <Card title="Edit Recipe">
          <Row>
            <Column width={2}>
              <Form.Label>Name:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.recipe.name}
                onChange={(event) => (this.recipe.name = event.currentTarget.value)}
              />
            </Column>
          </Row>

          <Row>
            <Column width={2}>Country:</Column>
            <Column>
              <Form.Select
                value={this.recipe.country}
                onChange={(event) => (this.recipe.country = event.currentTarget.value)}
              >
                {this.recipes
                  .map((recipe) => recipe.country)
                  .filter((country, index, array) => array.indexOf(country) === index)
                  .map((country, i) => (
                    <option key={i} value={country}>
                      {country}
                    </option>
                  ))}
              </Form.Select>
            </Column>
          </Row>

          <Row>
            <Column width={2}>
              <Form.Label>Category:</Form.Label>
            </Column>
            <Column>
              <Form.Select
                value={this.recipe.category}
                onChange={(event) => (this.recipe.category = event.currentTarget.value)}
              >
                {this.recipes
                  .map((recipe) => recipe.category)
                  .filter((category, index, array) => array.indexOf(category) === index)
                  .map((category, i) => (
                    <option key={i} value={category}>
                      {category}
                    </option>
                  ))}
              </Form.Select>
            </Column>
          </Row>
        </Card>
        <Row>
          <Column>
            <Button.Success
              onClick={() =>
                recipeService
                  .update(this.recipe)
                  .then(() => history.push('/recipes/' + this.recipe.recipe_id))
                  .catch((error) => Alert.danger('Error in updating recipe: ' + error.message))
              }
              //Tror problemet over her er at category og country blir lagt i recipes og ikke i recipe
            >
              Save
            </Button.Success>
          </Column>
          <Column right>
            <Button.Danger onClick={() => Alert.info('Not yet implemented')}>Delete</Button.Danger>
          </Column>
        </Row>
      </>
    );
  }

  mounted() {
    recipeService.getAll().then((recipes) => (this.recipes = recipes));

    recipeService
      .get(this.props.match.params.id)
      .then((recipe) => (this.recipe = recipe))
      .catch((error) => Alert.danger('Error getting recipe: ' + error.message));
  }
}
