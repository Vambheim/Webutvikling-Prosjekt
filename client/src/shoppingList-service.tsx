import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type ShoppingListInfo = {
  shopping_list_id: number;
  recipe_id: number;
  ingredient_id: number;
  name: string;
  amount: number;
  measurement_unit: string;
};

class ShoppingListService {
  /**
   * Get shopping list for given user
   */
  getShoppingList(user_id: number) {
    return axios
      .get<ShoppingListInfo[]>('/shoppinglist/' + user_id)
      .then((response) => response.data);
  }

  /**
   * Add to shopping list for given user
   */
  addToShoppingList(
    recipe_id: number,
    ingredient_id: number,
    user_id: number,
    amount: number,
    measurement_unit: string
  ) {
    return axios
      .post('/shoppinglist', {
        recipe_id: recipe_id,
        ingredient_id: ingredient_id,
        user_id: user_id,
        amount: amount,
        measurement_unit: measurement_unit,
      })
      .then((response) => response.data);
  }

  /**
   * Delete whole shopping list for given user
   */
  deleteShoppingList(user_id: number) {
    return axios.delete('/shoppinglist/' + user_id).then((response) => response.data);
  }

  /**
   * Delete one item in shopping list
   */
  deleteItemShoppingList(shopping_list_id: number) {
    return axios.delete('/shoppinglistitem/' + shopping_list_id).then((response) => response.data);
  }
}

const shoppingListService = new ShoppingListService();
export default shoppingListService;
