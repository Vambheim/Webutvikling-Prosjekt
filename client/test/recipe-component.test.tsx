import * as React from 'react';
import { Component } from 'react-simplified';
import { shallow } from 'enzyme';
import { Column, Alert } from '../src/widgets';
import { RecipeDetails, RecipeEdit, RecipeAdd, RecipeList } from '../src/recipe-components';
import recipeService, { Recipe } from '../src/recipe-service';
import { Form, Button, Card } from 'react-bootstrap';
import { NavLink } from 'react-router-dom';

jest.mock('../src/recipe-service', () => {
  class RecipeService {
    getAll() {
      return Promise.resolve([
        {
          id: 1098370,
          name: 'Butter Brie',
          category: 'side dish',
          country: 'American',
        },

        {
          id: 1098357,
          name: 'Three Ingredient Frozen Pina Colada',
          category: 'beverage',
          country: 'Mexican',
        },
        {
          id: 1098354,
          name: 'Rotisserie Chicken and Bean Tostadas',
          category: 'antipasti',
          country: 'Mexican',
        },
      ]);
    }
    create() {
      return Promise.resolve(4); // Same as: return new Promise((resolve) => resolve(4));
    }

    update() {
      return Promise.resolve();
    }

    delete() {
      return Promise.resolve();
    }

    get() {
      return Promise.resolve({
        id: 1098370,
        name: 'Butter Brie',
        category: 'side dish',
        country: 'American',
      });
    }

    getAllIngredients() {
      return Promise.resolve({
        ingredient_id: 1,
        name: 'carrot',
      });
    }
  }
  return new RecipeService();
});

describe('RecipeList tests', () => {
  test('RecipeList draws correctly', (done) => {
    const wrapper = shallow(<RecipeList />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Add filter button draws correctly ', (done) => {
    const wrapper = shallow(<RecipeList />);

    wrapper.find(Button).at(0).simulate('click');

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Add filter button draws correctly ', (done) => {
    const wrapper = shallow(<RecipeList />);

    wrapper.find(Button).at(1).simulate('click');

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Remove filters draws correctly', (done) => {
    const wrapper = shallow(<RecipeList />);

    wrapper.find(Button).at(2).simulate('click');

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Filter handles change', () => {
    const wrapper = shallow(<RecipeList />);

    wrapper
      .find(Form.Select)
      .at(0)
      .simulate('change', { currentTarget: { value: 'asian' } });
    wrapper
      .find(Form.Select)
      .at(1)
      .simulate('change', { currentTarget: { value: 'asian' } });
    wrapper
      .find(Form.Select)
      .at(2)
      .simulate('change', { currentTarget: { value: 'asian' } });
    wrapper
      .find(Form.Select)
      .at(3)
      .simulate('change', { currentTarget: { value: 'asian' } });
    wrapper
      .find(Form.Select)
      .at(4)
      .simulate('change', { currentTarget: { value: 'asian' } });

    expect(wrapper.containsMatchingElement(<Form.Control value="asian" />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<Form.Control value="asian" />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<Form.Control value="asian" />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<Form.Control value="asian" />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<Form.Control value="asian" />)).toEqual(false);
  });

  test('Searchbar handles change', () => {
    const wrapper = shallow(<RecipeList />);

    wrapper.find(Form.Control).simulate('change', { currentTarget: { value: 'Pannekaker' } });

    expect(wrapper.containsMatchingElement(<Form.Control value="Pannekaker" />)).toEqual(true);
  });
});

describe('RecipeDetails tests', () => {
  test('RecipeDetails draws correctly using snapshots', (done) => {
    const wrapper = shallow(<RecipeDetails match={{ params: { recipe_id: 1 } }} />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('RecipeDetails correctly directs to same page if logged in', (done) => {
    let loggedIn = false;
    const wrapper = shallow(<RecipeDetails match={{ params: { recipe_id: 1 } }} />);

    wrapper.find(Button).at(0).simulate('click');
    if (loggedIn) {
      setTimeout(() => {
        expect(location.hash).toEqual('#/recipes/:id');

        done();
      });
    } else {
      setTimeout(() => {
        expect(location.hash).toEqual('#/recipes/login');

        done();
      });
    }
  });

  test('Country and Category filter pushes correctly', (done) => {
    const wrapper = shallow(<RecipeDetails match={{ params: { recipe_id: 1 } }} />);
    wrapper.find(Button).at(0).simulate('click');
    setTimeout(() => {
      // $FlowExpectedError: do not type check next line.
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('RecipeDetails has 2 rows', () => {
    const wrapper = shallow(<RecipeDetails match={{ params: { recipe_id: 1 } }} />);

    expect(wrapper.find(Card)).toHaveLength(2);
  });

  test('Ingredientfilter pushes correctly', (done) => {
    const wrapper = shallow(<RecipeDetails match={{ params: { recipe_id: 1 } }} />);
    wrapper.find(Button).at(1).simulate('click');
    setTimeout(() => {
      // $FlowExpectedError: do not type check next line.
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
describe('RecipeAdd tests', () => {
  test('RecipeAdd draws correctly (using snapshot)', (done) => {
    const wrapper = shallow(<RecipeAdd />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('RecipeAdd correctly sets location save', (done) => {
    const wrapper = shallow(<RecipeAdd />);

    wrapper.find(Button).at(0).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/recipes/1/edit');

      done();
    });
  });

  test('RecipeAdd correctly sets location save', (done) => {
    const wrapper = shallow(<RecipeAdd />);

    wrapper.find(Button).at(4).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/recipes/1/edit');

      done();
    });
  });

  test(' RecipeAdd Form delivers change', () => {
    const wrapper = shallow(<RecipeAdd />);

    wrapper
      .find(Form.Control)
      .at(0)
      .simulate('change', { currentTarget: { value: 'milk' } });
    wrapper
      .find(Form.Control)
      .at(1)
      .simulate('change', { currentTarget: { value: 'milk' } });
    wrapper
      .find(Form.Control)
      .at(2)
      .simulate('change', { currentTarget: { value: 'milk' } });
    wrapper
      .find(Form.Control)
      .at(3)
      .simulate('change', { currentTarget: { value: 4 } });
    wrapper
      .find(Form.Control)
      .at(4)
      .simulate('change', { currentTarget: { value: 'milk' } });
    wrapper
      .find(Form.Control)
      .at(5)
      .simulate('change', { currentTarget: { value: 'milk' } });
    wrapper
      .find(Form.Control)
      .at(6)
      .simulate('change', { currentTarget: { value: 'milk' } });

    expect(wrapper.containsMatchingElement(<Form.Control value={'milk'} />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Control value={'milk'} />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Control value={'milk'} />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Control value={'milk'} />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Control value={'milk'} />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Control value={4} />)).toEqual(true);
  });
});

describe('RecipeEdit tests', () => {
  test('RecipeEdit draws correctly (using snapshot)', (done) => {
    const wrapper = shallow(<RecipeEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('EditRecipe correctly sets location save', (done) => {
    const wrapper = shallow(<RecipeEdit match={{ params: { id: 1 } }} />);

    wrapper.find(Button).at(1).simulate('click');

    expect(location.hash).toEqual('#/recipes/1/edit');

    done();
  });

  test('EditRecipe correctly sets location delete', (done) => {
    const wrapper = shallow(<RecipeEdit match={{ params: { id: 1 } }} />);

    wrapper.find(Button).at(0).simulate('click');

    expect(location.hash).toEqual('#/recipes/');

    done();
  });

  test('RecipeEdit delivers change', () => {
    const wrapper = shallow(<RecipeEdit match={{ params: { id: 1 } }} />);

    wrapper.find(Form.Control).simulate('change', { currentTarget: { value: 'Pannekaker' } });
    wrapper
      .find(Form.Select)
      .at(0)
      .simulate('change', { currentTarget: { value: 'Pannekaker' } });
    wrapper
      .find(Form.Select)
      .at(1)
      .simulate('change', { currentTarget: { value: 'Pannekaker' } });

    expect(wrapper.containsMatchingElement(<Form.Control value="Pannekaker" />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Select value="Pannekaker" />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Form.Select value="Pannekaker" />)).toEqual(true);
  });
});
