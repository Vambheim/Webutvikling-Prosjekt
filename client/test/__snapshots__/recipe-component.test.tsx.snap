// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDetails draws correctly (using snapshot) 1`] = `
<Fragment>
  Recipes:
</Fragment>
`;
